rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala

```diff
@@ -0,0 +1,334 @@
+
+      case (Nil, Vector(), Some(t)) => t :: Nil
+
+      case (Nil, Vector(), None) => EmptyTrie :: Nil
+    }
+
+  def process(actions: List[HistoryAction[K]]): History[K] = {
+    // TODO this is an intermediate step to reproduce all the trie behavior
+    // will evolve to a fold based implementation with partial tries
+    val newRoot = actions.foldLeft(this.root) {
+      case (currentRoot, InsertAction(k, value)) =>
+        val currentPath = asBytes(k)
+        val (_, path)   = traverseTrie(currentPath, currentRoot)
+        val elements    = rebalanceInsert(Leaf(value), currentPath, path)
+        historyStore.put(elements)
+        Trie.hash(elements.last)
+
+      case (currentRoot, DeleteAction(k)) =>
+        val currentPath   = asBytes(k)
+        val (_, triePath) = traverseTrie(currentPath, currentRoot)
+        val elements = triePath match {
+          case TriePath(elems :+ (_: Leaf), None, path) if currentPath == path =>
+            rebalanceDelete(currentPath, elems, None)
+          case _ =>
+            Nil
+        }
+        historyStore.put(elements)
+        elements.lastOption.map(Trie.hash).getOrElse(currentRoot)
+    }
+
+    this.copy(root = newRoot)
+  }
+
+  @SuppressWarnings(Array("org.wartremover.warts.Throw"))
+  private[history] def traverseTrie(key: KeyPath, start: Blake2b256Hash): (Trie, TriePath) = {
+    @tailrec
+    def traverse(t: Trie, currentPath: List[Byte], path: TriePath): (Trie, TriePath) =
+      (t, currentPath) match {
+        case (EmptyTrie, _) => (EmptyTrie, path)
+        case (l: Leaf, _)   => (t, path append (Nil, l))
+        case (s @ Skip(affix, p), _) if isPrefix(affix, currentPath) =>
+          traverse(historyStore.get(p),
+                   currentPath.drop(affix.size.toInt),
+                   path append (affix.toSeq.toList, s))
+        case (s: Skip, _) => //not a prefix
+          (EmptyTrie, path.copy(conflicting = Some(s)))
+        case (node @ Node(ptr), h :: tail) =>
+          pointerBlockStore.get(ptr) match {
+            case Some(pb) => traverse(pb.toVector(toInt(h)), tail, path append (h :: Nil, node))
+            case None     => throw new RuntimeException("malformed trie")
+          }
+        case _ => throw new RuntimeException("malformed trie")
+      }
+    traverse(historyStore.get(start), key, TriePath.empty)
+  }
+
+  private[history] def traverseTrie(key: KeyPath): (Trie, TriePath) =
+    traverseTrie(key, root)
+}
+
+object History {
+
+  def skipOrFetch(path: List[Byte], pointer: Blake2b256Hash, fetch: Blake2b256Hash => Trie): Trie =
+    if (path.isEmpty) {
+      fetch(pointer)
+    } else {
+      Skip(path.toByteVector, pointer)
+    }
+
+  def skipOrValue(path: List[Byte], t: Trie): Trie =
+    if (path.isEmpty) {
+      t
+    } else {
+      skip(t, path)
+    }
+
+  // this mapping is kept explicit on purpose
+  @inline
+  def toInt(b: Byte): Int = {
+    import java.lang.{Byte => JByte}
+    JByte.toUnsignedInt(b)
```

Comment:
`java.lang.Byte.toUnsignedInt(b)` is less cognitively demanding (no import, no rename, one line)